File: packages/protocol/src/manifest-proposal.ts
// A possible `armada.yml` per workspace, built from Scan, and the edits that
// iterate it. `crates/ipc/src/manifest_proposal.rs`.
//
// **Every line says where it came from, and nothing here sets that.** An edit
// carries no provenance; Fleet moves a touched line to `edited_during_setup` or
// `added_during_setup`. A port a file declares is `read`; every script is
// `convention`. Not `proposing.ts`, which is a Job on its way to the gate.
//
// **Lines, not text**: the file comes with Write, from `config`'s one writer.
//
// The header rules in `protocol.ts` hold here: hand-written, and every closed
// set is left as `string`.

/** `GET /repository/proposals`. */
export type ManifestProposals = {
  checkout: string;
  /** The root first, then by path — Scan's order. */
  proposals: ManifestProposal[];
  /** What a Job here stops at. **Stated, never written.** */
  caps: StatedCaps;
};

export type StatedCaps = { cost_micros: number; turns: number };

export type ManifestProposal = {
  /** The workspace, `.` for the root — what an edit names. */
  dir: string;
  /** Where the file would be, relative to the checkout. */
  file: string;
  /** The header, not a row. */
  id: ProposedId;
  ports: ProposedPort[];
  /** In written order, which is the order the gate starts them in. */
  checks: ProposedCheck[];
  commands: ProposedCommand[];
  /** Absent where nothing runs first. */
  setup?: ProposedSetup;
  /** `auto_merge` and `review_gate`, always both. */
  policy: ProposedPolicy[];
};

/**
 * `source` is `read`, `convention`, `default`, `edited_during_setup` or
 * `added_during_setup`. **Mono for a file, sans for the rest**: `file` and
 * `key` are present on the first two, and `key` is absent on a `convention`
 * whose whole file is the evidence.
 */
export type Provenance = { source: string; file?: string; key?: string };

export type ProposedId = { value: string; provenance: Provenance };

/** `env` is absent for a compose service, which gets its port without one. */
export type ProposedPort = {
  name: string;
  container?: number;
  env?: string;
  provenance: Provenance;
};

export type ProposedCheck = {
  name: string;
  run: string;
  /** Command names, in the order they run. Absent where none. */
  requires?: string[];
  provenance: Provenance;
};

export type ProposedCommand = {
  name: string;
  run: string;
  /** Absent where false. */
  destructive?: boolean;
  provenance: Provenance;
};

export type ProposedSetup = { requires: string[]; provenance: Provenance };

/** `key` is `auto_merge` or `review_gate`. A `default` row is an absent key. */
export type ProposedPolicy = { key: string; value: string; provenance: Provenance };

/** `POST /repository/edit_proposal`. */
export type EditManifestProposal = { dir: string; edit: ProposalEdit };

/**
 * A put naming a line that exists replaces it; one naming nothing adds it.
 * `band` is `ports`, `checks` or `commands`.
 */
export type ProposalEdit =
  | { edit: "id"; id: string }
  | { edit: "port"; name: string; container?: number; env?: string }
  | { edit: "check"; name: string; run: string; requires?: string[] }
  | { edit: "command"; name: string; run: string; destructive?: boolean }
  | { edit: "setup"; requires: string[] }
  | { edit: "policy"; key: string; value?: string }
  | { edit: "move"; name: string }
  | { edit: "remove"; band: string; name: string };

File: crates/ipc/src/manifest_proposal.rs
//! A possible `armada.yml` per workspace, built from Scan, as lines that each carry their
//! source. No edit names a provenance, so provenance stays a record.

use serde::{Deserialize, Serialize};

use crate::editing::ManifestSaved;
use crate::reading::ManifestRefused;

/// `get_manifest_proposals`: one proposal per workspace Scan found, root first.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ManifestProposals {
    pub checkout: String,
    pub proposals: Vec<ManifestProposal>,
    /// Stated at Setup and never written: no Job has run here to set them against.
    pub caps: StatedCaps,
}

/// What a Job stops at on this machine when its Manifest says nothing.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct StatedCaps {
    pub cost_micros: u64,
    pub turns: u64,
}

/// One workspace's proposal, as far as it has been iterated.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ManifestProposal {
    /// `.` for the root. What an edit names.
    pub dir: String,
    /// Where the file would be, relative to the checkout.
    pub file: String,
    pub id: ProposedId,
    pub ports: Vec<ProposedPort>,
    /// In written order, which is the order the gate starts them in.
    pub checks: Vec<ProposedCheck>,
    pub commands: Vec<ProposedCommand>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub setup: Option<ProposedSetup>,
    /// Always `auto_merge`, then `review_gate`.
    pub policy: Vec<ProposedPolicy>,
    /// The file Write would put down, from `config`'s writer. Absent where it is refused.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
    /// Every fault that stops the file loading. Write refuses while this is present.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub refused: Option<ManifestRefused>,
    /// Absent until Write lands it; after that the proposal takes no edits.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub written: Option<ManifestSaved>,
    /// An `armada.yml` is at `file` now — one already set up, or one Write put down.
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub present: bool,
}

/// Where a line came from.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "source", rename_all = "snake_case")]
pub enum Provenance {
    /// A file said this, at `key`.
    Read {
        file: String,
        key: String,
    },
    /// A file said this, and where it was placed is a guess. `key` is absent where the
    /// whole file is the evidence.
    Convention {
        file: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        key: Option<String>,
    },
    /// No file corresponds: the parser's value for an absent key, which Write leaves absent.
    Default,
    EditedDuringSetup,
    AddedDuringSetup,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ProposedId {
    pub value: String,
    pub provenance: Provenance,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ProposedPort {
    pub name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub container: Option<u16>,
    /// Absent for a compose service, which gets its port without one.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub env: Option<String>,
    pub provenance: Provenance,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ProposedCheck {
    pub name: String,
    pub run: String,
    /// Command names, in the order they run.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub requires: Vec<String>,
    /// Which runner drives this Check, where the scan's own evidence named
    /// one. **`None` is every other case** — a Check nothing detected a runner
    /// for runs whole, exactly as one in a repository nobody has configured
    /// does.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub runner: Option<ProposedRunner>,
    pub provenance: Provenance,
}

/// The runner one proposed Check names, and the package it runs in.
///
/// **Two fields and no commands.** Every way of running less than the whole
/// Check is written once in that runner's own description, never per Check —
/// `docs/concepts/runner-adapter.md`.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ProposedRunner {
    pub name: String,
    /// What `{pkg}` resolves to in that description's templates. **`None` on a
    /// workspace that is the repository root**, where there is no package
    /// below it to name.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub pkg: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ProposedCommand {
    pub name: String,
    pub run: String,
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub destructive: bool,
    pub provenance: Provenance,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ProposedSetup {
    pub requires: Vec<String>,
    pub provenance: Provenance,
}

/// `value` is in force either way. A `default` row is an absent key, so the file follows
/// the default if it moves.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ProposedPolicy {
    pub key: PolicyKey,
    pub value: String,
    pub provenance: Provenance,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PolicyKey {
    AutoMerge,
    ReviewGate,
}

/// `edit_manifest_proposal`: one change to one workspace's proposal.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct EditManifestProposal {
    pub dir: String,
    pub edit: ProposalEdit,
}

/// A put replaces the line it names or adds one. No variant carries provenance.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "edit", rename_all = "snake_case")]
pub enum ProposalEdit {
    Id {
        id: String,
    },
    Port {
        name: String,
        #[serde(default)]
        container: Option<u16>,
        #[serde(default)]
        env: Option<String>,
    },
    Check {
        name: String,
        run: String,
        #[serde(default)]
        requires: Vec<String>,
    },
    Command {
        name: String,
        run: String,
        #[serde(default)]
        destructive: bool,
    },
    /// Empty removes the row.
    Setup {
        requires: Vec<String>,
    },
    /// Absent `value` goes back to the default.
    Policy {
        key: PolicyKey,
        #[serde(default)]
        value: Option<String>,
    },
    /// A Check to the Commands, or a Command to the Checks.
    Move {
        name: String,
    },
    Remove {
        band: Band,
        name: String,
    },
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Band {
    Ports,
    Checks,
    Commands,
}

/// `write_manifest_proposal`: one workspace's proposal, put on disk as it stands.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct WriteManifestProposal {
    pub dir: String,
}
